### ISSUE TITLE
VS false positives.

### ISSUE THREAD

**Issue by eyalsk** [dotnet/roslyn#45017]
**Version Used**: 16.6.1

Sometimes analyzers provide false positives like in the following case:

![Unecessary assignment](https://user-images.githubusercontent.com/3756641/84218924-2219de80-aad8-11ea-8038-877327c8e613.png)

And sometimes I get red squiggles like in this case:

![Identifier expected](https://user-images.githubusercontent.com/3756641/84218954-31992780-aad8-11ea-80a4-d3d3591519ab.png)

The code compiles just fine.

I'm not sure why it happens so I can't provide any reproduction steps and I doubt these cases are related but maybe, dunno.

**Comment by CyrusNajmabadi** [comment]
> and I doubt these cases are related but maybe, dunno.

They are not related.

**Comment by CyrusNajmabadi** [comment]
> Sometimes analyzers provide false positives like in the following case:

This doesn't feel like a false positive.  But it's hard to tell because you haven't supplied your code.

The analyzer here is saying that the assignment of `false` is not necessary since htat value will never be used.  i.e. you could have assigned it to `true` and gotten the same result.

**Comment by eyalsk** [comment]
@CyrusNajmabadi The relevant [code](https://github.com/eyalsk/FindAndReplaceRobot/blob/2e4839094de05ebbd7e14c96c40dd47db51b2a11/src/FindAndReplaceRobot.Language/Lexer.cs#L98)

**Comment by CyrusNajmabadi** [comment]
The diagnostic is not saying that the variable isn't used.  It's saying that the *assigned value* is never used.   Looking at the code, the diagnostic is correct.

Because the loop is an infinite one, and because it can only exit after this line:

```c#
                    isError = !isNewLineOrEOF;
                    break;
```

the initial value assigned to `isError` is actually never read, and could be trivially changed to `true` (or removed) without affecting things.

**Comment by CyrusNajmabadi** [comment]
Went to gitter to chat about this more :)

**Comment by eyalsk** [comment]
@CyrusNajmabadi But then removing the assignment will lead to another error.

**Comment by eyalsk** [comment]
Sure.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi But then removing the assignment will lead to another error.

Can you show a screenshot of that?

**Comment by eyalsk** [comment]
Okay, it doesn't actually give me an error it changes it to `bool isError`.

**Comment by eyalsk** [comment]
I've discussed this with Cyrus and the analyzer is doing the right thing, all good so closing it.
